### PR TITLE
Facehugger behavior changes

### LINKED ISSUE #11920: Facehuggers can hug through mechs
#### Description of issue

While mining with a ripley, an unfortunate roboticist uncovered a xeno nest in the asteroid. Upon entering proximity to the nest, an egg opened

'Knochen Heinrich  says, "Dont tell me these things can go through the exo...."'

Said facehugger that left the egg pathed its way towards Knochen's mech, and took his mask off and facehugged him through an airtight mech suit.
#### Difference between expected and actual behavior

> Expected
> Get away from her you bitch
> Actual
> Get away from her you _facehugged_
#### Steps to reproduce

Be mining on asteroid with a RIPLEY
Find xeno nest
Get too close, so one of the eggs opens
Marvel that a hand with a penis inside of it is able to jump through a mech built for travelling through space
#### Server revision

713564a7711ee3d83f7a869ebd7ee786c64abf3a
#### Issue bingo

Please check whatever applies. More checkboxes checked increase your chances of not being yelled at by every contributor.

- [x] Issue could be reproduced at least once
- [x] Issue could be reproduced by different players
- [x] Issue could be reproduced in multiple rounds
- [x] Issue happened in a recent (less than 7 days ago) round
- [x] [Couldn't find an existing issue about this](https://github.com/d3athrow/vgstation13/issues)

### LINKED ISSUE #16924: Facehugger not Implanting.
(WEB REPORT BY: thewhargoul REMOTE: 172.93.109.202:7777)
# Revision
10af9cc89113dea8d5072ad2eb140e973866cdc2

# Description
If an alien places a hugger on someone manually by drag clicking and using the player's inventory menu to put the hugger as their mask, the hugger will not implant you.
# Steps to Reproduce
1. Be alien
2. put hugger on someone's mask slot by use of the menu
# What you Expected
To be implanted with an alien
# What Actually Happened
Didn't get implanted with an alien and was invulnerable from other huggers around me since one was already on my face.

### PR DESCRIPTION
**Pull request by Pathid**
Seen some complaints about big stacks of huggers being unfair to deal with. I'm submitting this as a potential resolution, if it's wanted that is. This PR gives them a very simple spreading out behavior: if a hugger that is on a turf and has no target finds another hugger on the same tile, it will move one turf in a ~~cardinal~~ direction.

In addition:
- ~~They'll drop off once they impregnate a host. Right now they die but stay on the face.~~  **_Nixed that, had another idea._**
- They'll go after corgis, which they are able to impregnate already but have to be manually applied for some reason.
(More added since. See below.)

Closes #11920
Closes #16924

:cl:
* rscadd: Facehuggers now wander a little if they find another facehugger on their turf.
* bugfix: Toy facehuggers no longer hunt or remove equipment (although Lamarr still will). Wearing a dead, toy, or sterile facehugger no longer fools hugger AI. Huggers should no longer hug through mechs, closets or other containers.
* tweak: Facehuggers will try to keep their victims alive by auto-connecting to any internals they had enabled. (Don't ask how they know how to do that.)

### CODE REVIEW

**Comment by Dacendeth** [comment]
does this mean if a bunch are in a locker they'll attempt to walk out of it?

**Comment by Pathid** [comment]
@Dacendeth If it's open, yes. It would be possible to fill up a locker, but harder.

**Comment by CptWad** [comment]
fair enough
and this would also mean that if you did fill up a locker it would be 10 time more terrifying as 20 something huggers proceeded to flood in all directions once opening pandoras box of facerape

File: code/modules/mob/living/carbon/alien/special/facehugger.dm [hunk 1]
```diff
@@ -95,6 +96,13 @@
 					findtarget()
 					return
 
+/obj/item/clothing/mask/facehugger/proc/spreadout()
+	if(!target && isturf(loc))
+		for(var/obj/item/clothing/mask/facehugger/F in loc)
+			if(F != src)
+				step(src, pick(1,2,4,8), 0)
```
Comment (Kurfursten) [inline code/modules/mob/living/carbon/alien/special/facehugger.dm]: we have global list `cardinal` for this!

it's functionally the same but reads nicer, like defines

**Comment by gbasood** [comment]
neat
I'm honestly of the opinion that non-toy, non-Lamarr facehuggers just shouldnt be able to be in containers at all but this is also good and should happen regardless

**Comment by Pathid** [comment]
Speaking of the toy huggers, whoever put in the checks for the ``real`` var clearly intended the toy versions not to hunt or rip off equipment, but nothing sets ``real`` to false in their case, or any case.

**Comment by gbasood** [comment]
@Pathid the subtype for toy and lamarr should be doing that

**Comment by Pathid** [comment]
So apparently the corgi hugging needs work, and not only because it says `mob/living/carbon` on line 60. Also I'm going update it so hugging through mechs and closets doesn't happen anymore.

**Comment by Pathid** [comment]
Changelog updated.

File: code/modules/mob/living/carbon/alien/special/facehugger.dm [hunk 2]
```diff
@@ -287,10 +293,38 @@
...
+			else if(istype(headwear, /obj/item/clothing/head/helmet/space/rig))
+				rng = CHANCE_TO_REMOVE_SPECIAL_HEADWEAR
+
+			if(prob(rng))
+				C.visible_message("<span class='danger'>\The [src] smashes against [C]'s \the [headwear], and rips it off in the process!</span>")
+				C.drop_from_inventory(headwear)
+				GoIdle(TIME_IDLE_AFTER_HEAD_DENIED)
+				return
+			else
+				C.visible_message("<span class='danger'>\The [src] bounces off of the [headwear]!</span>")
```
Comment (DamianX) [inline code/modules/mob/living/carbon/alien/special/facehugger.dm]: \the

Comment (Pathid) [reply code/modules/mob/living/carbon/alien/special/facehugger.dm]: So \The is redundant here, or is this a case-sensitivity thing?

Comment (DamianX) [reply code/modules/mob/living/carbon/alien/special/facehugger.dm]: `the [headwear]` will print `the the security helmet`
make it `\the [headwear]`

Comment (Pathid) [reply code/modules/mob/living/carbon/alien/special/facehugger.dm]: Whoops

**Comment by Pathid** [comment]
`[14:44:21] Runtime in facehugger.dm,401: undefined variable /obj/item/clothing/mask/facehugger/var/status_flags`

Also trying to figure out how to fix #16924

File: code/modules/mob/living/carbon/alien/special/facehugger.dm [hunk 3]
```diff
@@ -52,16 +52,16 @@
...
-	for(var/mob/living/carbon/T in hearers(src,4))
-		if(!ishuman(T) && !ismonkey(T))
-			continue
-		if(!CanHug(T))
+	target = null
+	for(var/mob/living/T in hearers(src,4))
+		if(!CanHug(T, src))
 			continue
-		if(T && (T.stat != DEAD && T.stat != UNCONSCIOUS) )
+		if(T && (!T.isDead() && !T.isUnconscious() ) )
```
Comment (DamianX) [inline code/modules/mob/living/carbon/alien/special/facehugger.dm]: `isDead` is redundant if you're checking for `isUnconscious`

File: code/modules/mob/living/carbon/alien/special/facehugger.dm [hunk 4]
```diff
@@ -70,31 +70,38 @@
 /obj/item/clothing/mask/facehugger/proc/followtarget()
 	if(!real)
 		return // Why are you trying to path stupid toy
-	if(!target || target.stat == DEAD || target.stat == UNCONSCIOUS || target.status_flags & XENO_HOST)
+	if(!target || target.isDead() || target.isUnconscious() || target.status_flags & XENO_HOST)
```
Comment (DamianX) [inline code/modules/mob/living/carbon/alien/special/facehugger.dm]: `isDead` is redundant if you're checking for `isUnconscious`

File: code/modules/mob/living/carbon/alien/special/facehugger.dm [hunk 5]
```diff
@@ -192,8 +199,10 @@
 	return FALSE
 
 /obj/item/clothing/mask/facehugger/HasProximity(atom/movable/AM as mob|obj)
-	if(CanHug(AM))
-		return Attach(AM)
+	if(isliving(AM))
+		var/mob/living/L = AM
```
Comment (DamianX) [inline code/modules/mob/living/carbon/alien/special/facehugger.dm]: superfluous cast, you can get rid of it

**Comment by Probe1** [comment]
Combine your bugfix changelog stuff.

**Pull request merged**